contracts/SpokePoolV3Periphery.sol

```diff
@@ -53,153 +84,153 @@ contract SpokePoolPeripheryProxy is Lockable, MultiCaller {
         SpokePoolV3Periphery.DepositData calldata depositData,
         uint256 deadline,
-        uint8 v,
-        bytes32 r,
-        bytes32 s
+        bytes calldata permitSignature
     ) external nonReentrant {
+        (bytes32 r, bytes32 s, uint8 v) = _deserializeSignature(permitSignature);
         // For permit transactions, we wrap the call in a try/catch block so that the transaction will continue even if the call to
         // permit fails. For example, this may be useful if the permit signature, which can be redeemed by anyone, is executed by somebody
         // other than this contract.
-        try swapToken.permit(msg.sender, address(this), swapTokenAmount, deadline, v, r, s) {} catch {}
-        _callSwapAndBridge(
-            IERC20(address(swapToken)),
-            acrossInputToken,
-            exchange,
-            routerCalldata,
-            swapTokenAmount,
-            minExpectedInputTokenAmount,
-            depositData
+        try
+            IERC20Permit(address(swapData.swapToken)).permit(
+                msg.sender,
+                address(this),
+                swapData.swapTokenAmount,
+                deadline,
+                v,
+                r,
+                s
+            )
+        {} catch {}
+        _callSwapAndBridge(acrossInputToken, swapData, depositData);
+    }
+
+    /**
+     * @notice Uses permit2 to transfer tokens from a user before swapping a token on this chain via specified router and submitting an Across deposit atomically.
+     * Caller can specify their slippage tolerance for the swap and Across deposit params.
+     * @dev This function assumes the caller has properly set an allowance for the permit2 contract on this network.
+     * @dev This function assumes that the amount of token to be swapped is equal to the amount of the token to be received from permit2.
+     * @param acrossInputToken Address of the token that will be bridged via Across as the inputToken.
+     * @param depositor The owner of the permit2 signature and depositor for the Across spoke pool.
+     * @param swapData Specifies the params we need to perform a swap on a generic exchange.
+     * @param depositData Specifies the Across deposit params we'll send after the swap.
+     * @param permit The permit data signed over by the owner.
+     * @param transferDetails The spender's requested transfer details for the permitted token.
+     * @param signature The permit2 signature to verify against the deposit data.
+     */
+    function swapAndBridgeWithPermit2(
+        IERC20 acrossInputToken,
+        address depositor,
+        SpokePoolV3Periphery.SwapData calldata swapData,
+        SpokePoolV3Periphery.DepositData calldata depositData,
+        IPermit2.PermitTransferFrom calldata permit,
+        IPermit2.SignatureTransferDetails calldata transferDetails,
+        bytes calldata signature
+    ) external nonReentrant {
+        bytes32 witness = keccak256(
```

Comment:
Permit2 cannot be approval abused, so there should be no need for an extra hop as long as we include _all_ relevant swap and deposit data in the permit2 signed witness data.

Should we consider only using the extra hop for approvals, permit, and transferWithAuthorization?

Reply:
Bennett and I discussed this as well. We discussed moving swapAndBridgeWithPermit2 to the Periphery in a separate PR but since this PR only adds Permit2 functionality we should be good to put all that permit2 logic in the Periphery rather than the Proxy. @bmzig thoughts?

Reply:
Yeah I think this works. I agree that we will need to refactor this in general once the correct functionality to the non-approval based flows is finished.

Reply:
I've moved the permit and auth functions back to the periphery, which reduced the code diff a lot 

https://github.com/across-protocol/contracts/pull/777/files